test/e2e/prometheus_test.go

```diff
@@ -1938,6 +1938,151 @@ func testPromSharedResourcesReconciliation(t *testing.T) {
 	}
 }
 
+func testShardingProvisioning(t *testing.T) {
+	t.Parallel()
+
+	ctx := framework.NewTestCtx(t)
+	defer ctx.Cleanup(t)
+	ns := ctx.CreateNamespace(t, framework.KubeClient)
+	ctx.SetupPrometheusRBAC(t, ns, framework.KubeClient)
+
+	prometheusName := "test"
+	group := "servicediscovery-test"
+	svc := framework.MakePrometheusService(prometheusName, group, v1.ServiceTypeClusterIP)
+
+	s := framework.MakeBasicServiceMonitor(group)
+	if _, err := framework.MonClientV1.ServiceMonitors(ns).Create(context.TODO(), s, metav1.CreateOptions{}); err != nil {
+		t.Fatal("Creating ServiceMonitor failed: ", err)
+	}
+
+	p := framework.MakeBasicPrometheus(ns, prometheusName, group, 1)
+	shards := int32(2)
+	p.Spec.Shards = &shards
+	p, err := framework.CreatePrometheusAndWaitUntilReady(ns, p)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if finalizerFn, err := testFramework.CreateServiceAndWaitUntilReady(framework.KubeClient, ns, svc); err != nil {
+		t.Fatal(errors.Wrap(err, "creating prometheus service failed"))
+	} else {
+		ctx.AddFinalizerFn(finalizerFn)
+	}
+
+	pods := []struct {
+		pod                        string
+		expectedShardConfigSnippet string
+	}{
+		{
+			pod: "prometheus-test-0",
+			expectedShardConfigSnippet: `
+  - source_labels:
+    - __tmp_hash
+    regex: 0
+    action: keep`,
+		}, {
+			pod: "prometheus-test-shard-1-0",
+			expectedShardConfigSnippet: `
+  - source_labels:
+    - __tmp_hash
+    regex: 1
+    action: keep`,
+		},
+	}
+
+	for _, p := range pods {
+		stdout, _, err := framework.ExecWithOptions(testFramework.ExecOptions{
+			Command: []string{
+				"/bin/sh", "-c", "cat /etc/prometheus/config_out/prometheus.env.yaml",
+			},
+			Namespace:     ns,
+			PodName:       p.pod,
+			ContainerName: "prometheus",
+			CaptureStdout: true,
+			CaptureStderr: true,
+			Stdin:         nil,
+		})
+		if err != nil {
+			t.Fatalf("Failed to read config from pod %q: %v", p.pod, err)
+		}
+		if !strings.Contains(stdout, p.expectedShardConfigSnippet) {
```

Comment:
Is this e2e test enough to test the sharding feature, my only worry here is we don't test the sharding feature just Prometheus configuration and even that not in full just that the parts of config are there. Is there another way we could do an e2e test?

Reply:
1) Sharding is already tested in Prometheus, it feels off to double test it here.
2) We shard on the `__address__` label, which is a `Pod`s IP, which is not something we can deterministically implement in an e2e test, unfortunately.